breakpad-patches/16-get-last-error.patch

```diff
@@ -0,0 +1,181 @@
+diff --git a/src/google_breakpad/processor/call_stack.h b/src/google_breakpad/processor/call_stack.h
+--- a/src/google_breakpad/processor/call_stack.h
++++ b/src/google_breakpad/processor/call_stack.h
+@@ -62,26 +62,30 @@ class CallStack {
+ 
+   // Resets the CallStack to its initial empty state
+   void Clear();
+ 
+   const vector<StackFrame*>* frames() const { return &frames_; }
+ 
+   // Set the TID associated with this call stack.
+   void set_tid(uint32_t tid) { tid_ = tid; }
++  void set_last_error(uint32_t last_error) { last_error_ = last_error; }
+ 
+   uint32_t tid() const { return tid_; }
++  uint32_t last_error() const { return last_error_; }
+ 
+  private:
+   // Stackwalker is responsible for building the frames_ vector.
+   friend class Stackwalker;
+ 
+   // Storage for pushed frames.
+   vector<StackFrame*> frames_;
+ 
+   // The TID associated with this call stack. Default to 0 if it's not
+   // available.
+   uint32_t tid_;
++  // The last error the OS set for this thread (win32's GetLastError())
++  uint32_t last_error_;
+ };
+ 
+ }  // namespace google_breakpad
+ 
+ #endif  // GOOGLE_BREAKPAD_PROCSSOR_CALL_STACK_H__
+diff --git a/src/google_breakpad/processor/minidump.h b/src/google_breakpad/processor/minidump.h
+--- a/src/google_breakpad/processor/minidump.h
++++ b/src/google_breakpad/processor/minidump.h
+@@ -279,16 +279,26 @@ class MinidumpMemoryRegion : public Mini
+ class MinidumpThread : public MinidumpObject {
+  public:
+   virtual ~MinidumpThread();
+ 
+   const MDRawThread* thread() const { return valid_ ? &thread_ : NULL; }
+   // GetMemory may return NULL even if the MinidumpThread is valid,
+   // if the thread memory cannot be read.
+   virtual MinidumpMemoryRegion* GetMemory();
++  // Corresponds to win32's GetLastError function, which records the last
++  // error value set by the OS for this thread. A more useful error message
++  // can be produced by passing this value to FormatMessage:
++  //
++  // https://docs.microsoft.com/windows/win32/debug/retrieving-the-last-error-code
++  //
++  // The value may also be looked up in Microsoft's System Error Codes listing:
++  //
++  // https://docs.microsoft.com/windows/win32/debug/system-error-codes
++  virtual uint32_t GetLastError();
+   // GetContext may return NULL even if the MinidumpThread is valid.
+   virtual MinidumpContext* GetContext();
+ 
+   // The thread ID is used to determine if a thread is the exception thread,
+   // so a special getter is provided to retrieve this data from the
+   // MDRawThread structure.  Returns false if the thread ID cannot be
+   // determined.
+   virtual bool GetThreadID(uint32_t *thread_id) const;
+diff --git a/src/processor/call_stack.cc b/src/processor/call_stack.cc
+--- a/src/processor/call_stack.cc
++++ b/src/processor/call_stack.cc
+@@ -44,11 +44,12 @@ CallStack::~CallStack() {
+ 
+ void CallStack::Clear() {
+   for (vector<StackFrame *>::const_iterator iterator = frames_.begin();
```

Comment:
This line reads:
```
   for (vector<StackFrame*>::const_iterator iterator = frames_.begin();
```
In Breakpad sources, I think this is why the patch doesn't apply.